### PR TITLE
3046 csv import nested thesauri

### LINKED ISSUE #3046: Allow groupings on CSV Import of thesauri
Since we allow grouping or 'nesting' of thesaurus values, we should implement a way to achieve that structure via CSV Import.

### PR DESCRIPTION
**Pull request by fnocetti**
fixes #3046 

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

File: app/api/csv/specs/csvLoaderThesauri.spec.js [hunk 1]
```diff
@@ -118,5 +118,136 @@ describe('csvLoader thesauri', () => {
       ]);
       mockedFile.mockRestore();
     });
+
+    describe('nesting', () => {
```
Comment (LaszloKecskes) [inline app/api/csv/specs/csvLoaderThesauri.spec.js:121]: Please consider adding one test case, which checks the value of translations.

File: app/api/thesauri/thesauri.js [hunk 2]
```diff
@@ -135,6 +135,22 @@ const update = async thesauri => {
...
 
+function recursivelyAppendValues(originalValues, newValues) {
+  const values = [...originalValues];
+  const existingLabels = new Set(values.map(v => v.label));
+
+  newValues.forEach(newValue => {
+    if (!existingLabels.has(newValue.label)) {
+      values.push(newValue);
+    } else if (newValue.values) {
+      const originalValue = values.find(v => v.label === newValue.label);
```
Comment (LaszloKecskes) [inline app/api/thesauri/thesauri.js]: Find inside a foreach. I don't know how big are our thesauri's, but I have been surprised the past. Consider mapping the original labels by value before the forEach.

File: app/api/csv/importThesauri.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import { LanguageSchema } from 'shared/types/commonTypes';
+import { ThesaurusValueSchema } from 'shared/types/thesaurusType';
+import { Readable } from 'stream';
+import { CSVRow } from './csv';
+
+type ParsedValue = { nested: boolean; value: string };
+type ParsedRow = Record<string, ParsedValue>;
+
+const buildThesauri = (rows: ParsedRow[], languageLabel: string) =>
+  rows.reduce<ThesaurusValueSchema[]>((thesauriValues, row) => {
```
Comment (LaszloKecskes) [inline app/api/csv/importThesauri.ts]: Using reduce to build an array. The same concern as before about the thesaurus length we might have. Consider using a different approach. This seems complicated, so probably a loop.

Also, the same concern about using concat: it also copies every existing thing again, to return a new array. After resolving the loop, consider using an in-place modification, like push.

**Review by LaszloKecskes — CHANGES_REQUESTED** [review]
The functionality works great, but there is some discussion needed about the implementation.

File: app/api/csv/importThesauri.ts (hunk 3 above)
Comment (fnocetti) [reply app/api/csv/importThesauri.ts]: makes sense

File: app/api/csv/specs/csvLoaderThesauri.spec.js (hunk 1 above)
Comment (fnocetti) [reply app/api/csv/specs/csvLoaderThesauri.spec.js:121]: sure

File: app/api/thesauri/thesauri.js (hunk 2 above)
Comment (fnocetti) [reply app/api/thesauri/thesauri.js]: makes sense as well

File: app/api/csv/specs/csvLoaderThesauri.spec.js (hunk 1 above)
Comment (fnocetti) [reply app/api/csv/specs/csvLoaderThesauri.spec.js:121]: done

File: app/api/thesauri/thesauri.js (hunk 2 above)
Comment (fnocetti) [reply app/api/thesauri/thesauri.js]: done

File: app/api/csv/importThesauri.ts (hunk 3 above)
Comment (fnocetti) [reply app/api/csv/importThesauri.ts]: done

**Review by LaszloKecskes — APPROVED** [review]
Every request is addressed, perfect. Thanks!

**Pull request merged**